src/gt4py/next/program_processors/runners/dace_iterator/__init__.py

```diff
@@ -44,13 +44,19 @@ def get_sorted_dim_ranges(domain: Domain) -> Sequence[UnitRange]:
 )
 
 
-def convert_arg(arg: Any):
+def convert_arg(arg: Any, run_on_gpu: bool):
     if is_field(arg):
         sorted_dims = get_sorted_dims(arg.domain.dims)
         ndim = len(sorted_dims)
         dim_indices = [dim_index for dim_index, _ in sorted_dims]
-        assert isinstance(arg.ndarray, np.ndarray)
-        return np.moveaxis(arg.ndarray, range(ndim), dim_indices)
+        if run_on_gpu:
+            import cupy as cp
+
+            assert isinstance(arg.ndarray, cp.ndarray)
+            return cp.moveaxis(arg.ndarray, range(ndim), dim_indices)
+        else:
+            assert isinstance(arg.ndarray, np.ndarray)
+            return np.moveaxis(arg.ndarray, range(ndim), dim_indices)
```

Comment:
I think it's better to keep `convert_arg` independent of the hardware acceleration. Could you rewrite `convert_arg` in a way such that it no longer expects a `run_on_gpu` flag, but rather just simply recognizes the argument and converts it accordingly? You can first check if it's a numpy array, and if it not, you can import cupy in a `try` block, and handle the `ImportError` or return the processed cupy array. This would save you from passing the `run_on_gpu` flag down the entire call stack.

Reply:
Good idea. I have also elaborated it a bit more.